File: src/main/java/com/mattchowning/file_read_write/server/handler/ServerFileReadWriteHandler.java
package com.mattchowning.file_read_write.server.handler;

import com.mattchowning.file_read_write.server.ServerUtil;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Base64;

import javax.activation.MimetypesFileTypeMap;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.*;
import io.netty.handler.codec.http.*;

public class ServerFileReadWriteHandler extends SimpleChannelInboundHandler<FullHttpRequest> {

    private final FileManager fileManager;
    private final ServerUtil serverUtil;

    public ServerFileReadWriteHandler(String fullFilePath) {
        this(new FileManager(fullFilePath), new ServerUtil());
    }

    protected ServerFileReadWriteHandler(FileManager fileManager, ServerUtil serverUtil) {
        super();
        this.fileManager = fileManager;
        this.serverUtil = serverUtil;
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
        switch (request.method().toString()) {
            case "GET":
                if (isClientAuthorized(ctx)) {
                    returnFileContent(ctx);
                } else {
                    returnEncodedFileContent(ctx);
                }
                break;
            case "POST":
                if (isClientAuthorized(ctx)) {
                    writeFileContent(ctx, request.content());
                } else {
                    serverUtil.sendError(ctx,
                                         HttpResponseStatus.UNAUTHORIZED,
                                         "invalid_request",
                                         "Authentication required to update file.");
                }
                break;
            default:
                serverUtil.sendError(ctx, HttpResponseStatus.METHOD_NOT_ALLOWED);
        }
    }

    private boolean isClientAuthorized(ChannelHandlerContext ctx) {
        return ctx.channel().hasAttr(ServerOAuthVerificationHandler.AUTHORIZED) &&
               ctx.channel().attr(ServerOAuthVerificationHandler.AUTHORIZED).get();
    }

    private void returnEncodedFileContent(ChannelHandlerContext ctx) {
        try {
            File file = fileManager.getFile();
            byte[] bytes = getBase64EncodedFileBytes(file);
            HttpResponse response = getHttpResponse(bytes);
            ctx.writeAndFlush(response);
        } catch (IOException e) {
            serverUtil.sendError(ctx, HttpResponseStatus.CONFLICT);
            e.printStackTrace();
        }
    }

    private byte[] getBase64EncodedFileBytes(File file) throws IOException {
        byte[] bytes = Files.readAllBytes(file.toPath());
        return Base64.getEncoder().encode(bytes);
    }

    private void returnFileContent(ChannelHandlerContext ctx) {
        File file = fileManager.getFile();
        HttpResponse response = getHttpResponse(file);
        DefaultFileRegion fileContent = new DefaultFileRegion(file, 0, file.length());
        ctx.write(response);
        ctx.writeAndFlush(fileContent);
    }

    @NotNull
    private HttpResponse getHttpResponse(File file) {
        HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
        response.headers().set(HttpHeaderNames.CONTENT_LENGTH, file.length());
        response.headers().set(HttpHeaderNames.CONTENT_TYPE, getContentType(file));
        response.headers().set(HttpHeaderNames.DATE, serverUtil.getDate());
        return response;
    }

    @NotNull
    private HttpResponse getHttpResponse(byte[] bytes) {
        FullHttpResponse response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
                                                                HttpResponseStatus.OK,
                                                                Unpooled.copiedBuffer(bytes));
        response.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain");
        response.headers().set(HttpHeaderNames.DATE, serverUtil.getDate());
        response.headers().set(HttpHeaderNames.CONTENT_LENGTH, bytes.length);
        return response;
    }

    private void writeFileContent(ChannelHandlerContext ctx, ByteBuf content) {
        boolean succeeded = fileManager.writeFileContent(content);
        if (succeeded) {
            returnFileContent(ctx);
        } else {
            serverUtil.sendError(ctx, HttpResponseStatus.CONFLICT);
        }
    }

    private String getContentType(File file) {
        return new MimetypesFileTypeMap().getContentType(file.getPath());
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        cause.printStackTrace();
        ctx.close();
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/server/handler/ServerFileReadWriteHandler.scala
package com.mattchowning_scala.file_read_write.server.handler

import com.mattchowning_scala.file_read_write.server.ServerUtil
import java.io.File
import javax.activation.MimetypesFileTypeMap
import io.netty.buffer.ByteBuf
import io.netty.channel._
import io.netty.handler.codec.http._

class ServerFileReadWriteHandler(fileManager: FileManager, serverUtil: ServerUtil)
  extends SimpleChannelInboundHandler[FullHttpRequest] {

  def this(fullFilePath: String) {
    this(new FileManager(fullFilePath), new ServerUtil)
  }

  @throws(classOf[Exception])
  protected def channelRead0(ctx: ChannelHandlerContext, request: FullHttpRequest) {
    request.method.toString match {
      case "GET"  => returnFileContent(ctx)
      case "POST" => writeFileContent(ctx, request.content)
      case _      => serverUtil.sendError(ctx, HttpResponseStatus.METHOD_NOT_ALLOWED)
    }
  }

  private def returnFileContent(ctx: ChannelHandlerContext) {
    val file: File = fileManager.getFile
    val response: HttpResponse = getHttpResponse(file)
    val fileContent: DefaultFileRegion = new DefaultFileRegion(file, 0, file.length)
    ctx.write(response)
    ctx.writeAndFlush(fileContent)
  }

  private def getHttpResponse(file: File): HttpResponse = {
    val response: HttpResponse = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK)
    response.headers.set(HttpHeaderNames.CONTENT_LENGTH, file.length)
    response.headers.set(HttpHeaderNames.CONTENT_TYPE, getContentType(file))
    response.headers.set(HttpHeaderNames.DATE, serverUtil.getDate)
    response
  }

  private def writeFileContent(ctx: ChannelHandlerContext, content: ByteBuf) {
    val succeeded: Boolean = fileManager.writeFileContent(content)
    if (succeeded) {
      returnFileContent(ctx)
    } else {
      serverUtil.sendError(ctx, HttpResponseStatus.NO_CONTENT)
    }
  }

  private def getContentType(file: File): String = {
    new MimetypesFileTypeMap().getContentType(file.getPath)
  }

  @throws(classOf[Exception])
  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
    cause.printStackTrace()
    ctx.close
  }
}